pkg/velero/server/args.go

```diff
@@ -47,7 +48,7 @@ func (a Args) StringArr(dpaFeatureFlags []string, logLevel string) ([]string, er
 		args = append(args, fmt.Sprintf("--log-level=%s", logrusLevel.String()))
 	}
 	if a.BackupSyncPeriod != nil {
-		args = append(args, fmt.Sprintf("--backup-sync-period=%s", a.BackupSyncPeriod.String())) // duration
+		args = append(args, fmt.Sprintf("--backup-sync-period=%s", (time.Duration(*a.BackupSyncPeriod)*time.Minute).String())) // duration
```

Comment:
This changes the field behavior, but I do not know if anyone ever used them. Without this, to pass a minute, user would input `60000000000` right? Prior to making this changing to all fields, I want the approval of the team

Reply:
Checked and all these time fields are in nano seconds

```
args:
  - server
  - '--backup-sync-period=1ns'
  - '--default-backup-ttl=1ns'
  - '--default-item-operation-timeout=1ns'
  - '--resource-timeout=1ns'
  - '--default-repo-maintain-frequency=1ns'
  - '--garbage-collection-frequency=1ns'
  - '--fs-backup-timeout=1ns'
  - '--item-operation-sync-frequency=1ns'
  - >-
    --restore-resource-priorities=securitycontextconstraints,customresourcedefinitions,namespaces,managedcluster.cluster.open-cluster-management.io,managedcluster.clusterview.open-cluster-management.io,klusterletaddonconfig.agent.open-cluster-management.io,managedclusteraddon.addon.open-cluster-management.io,storageclasses,volumesnapshotclass.snapshot.storage.k8s.io,volumesnapshotcontents.snapshot.storage.k8s.io,volumesnapshots.snapshot.storage.k8s.io,datauploads.velero.io,persistentvolumes,persistentvolumeclaims,serviceaccounts,secrets,configmaps,limitranges,pods,replicasets.apps,clusterclasses.cluster.x-k8s.io,endpoints,services,-,clusterbootstraps.run.tanzu.vmware.com,clusters.cluster.x-k8s.io,clusterresourcesets.addons.cluster.x-k8s.io
  - '--store-validation-frequency=1ns'
  - '--terminating-resource-timeout=1ns'
```  

My question is: should I update the unit of measure of this fields?

Reply:
> Without this, to pass a minute, user would input 60000000000 right?

No. User would put into their yaml `4h` for example.
```
	// How long pod volume file system backups/restores should be allowed to run before timing out. (default 4h0m0s)
	// +optional
	PodVolumeOperationTimeout *time.Duration `json:"fs-backup-timeout,omitempty"`
```

Per the example.

Somewhere there is a text during parsing from string into time.Duration for you.

Reply:
The unmarshal function https://pkg.go.dev/time#Time.UnmarshalJSON for time package will be used during yaml unmarshalling via https://pkg.go.dev/sigs.k8s.io/yaml

> In short, this library first converts YAML to JSON using go-yaml and then uses json.Marshal and json.Unmarshal to convert to or from the struct. This means that it effectively reuses the JSON struct tags as well as the custom JSON methods MarshalJSON and UnmarshalJSON unlike go-yaml. For a detailed overview of the rationale behind this method, [see this blog post](http://web.archive.org/web/20190603050330/http://ghodss.com/2014/the-right-way-to-handle-yaml-in-golang/).

I would not make this change.

Reply:
time.Duration.UnmarshalJSON function is used and can accept string in the yaml.
> UnmarshalJSON implements the json.Unmarshaler interface. The time must be a quoted string in the [RFC 3339](https://rfc-editor.org/rfc/rfc3339.html) format.

Reply:
![Screenshot from 2024-02-01 14-57-02](https://github.com/openshift/oadp-operator/assets/66965232/19db6b89-1b6c-488c-8b85-d6bc247b797f)

Its not possible (1.3 example), I can not even create the DPA using hour syntax

Reply:
Ok I can't find the specific place where this is implemented.. but point is it can decode duration.

Here's the unit test https://github.com/kubernetes-sigs/yaml/blob/c3772b51db126345efe2dfe4ff8dac83b8141684/goyaml.v3/decode_test.go#L926-L931
```
func (s *S) TestUnmarshalDurationInt(c *C) {
	// Don't accept plain ints as durations as it's unclear (issue #200).
	var d time.Duration
	err := yaml.Unmarshal([]byte("123"), &d)
	c.Assert(err, ErrorMatches, "(?s).* line 1: cannot unmarshal !!int `123` into time.Duration")
}
```

Reply:
How would we change this validation without changing the CRD?

Here an example which accepts time syntax https://github.com/openshift/oadp-operator/blob/master/api/v1alpha1/oadp_types.go#L202 https://github.com/openshift/oadp-operator/blob/master/bundle/manifests/oadp.openshift.io_dataprotectionapplications.yaml#L46-L49 but to work like this, CRD change would be necessary, no?

Reply:
If it really was broken and no one could use it then CRD change would be ok if we can assume no one used it.

Reply:
I think so, I do not see people passing nano seconds to these fields

I believe there are 2 options:

- just update CRD descriptions telling the unit of measure (and then, without changing CRD, we can change the unit of measure from nano seconds, to minutes, for example)
- (CHANGES CRD) fix field to accept time syntax string (better, but more dangerous)

Reply:
Ok I found the issue.. have to convert all the time.Duration to metav1.Duration. Then all is good.

Reply:
that works, but would change the CRD (type integer to string) (so follows to second option)

Reply:
PR: https://github.com/openshift/oadp-operator/pull/1306

Reply:
resolving as we have agreed on just fix fields descriptions, without changing their unit of measure